### ISSUE TITLE
Progress spinner shakes when specifying a diameter.

### ISSUE THREAD

**Issue by HeathJared** [angular/material2#11756]
#### Bug:

When specifying a diameter (16-32px) on the progress spinner the animation will eventually start to shake by a couple pixels as it animates. Using smaller diameters make it more noticeable. Using other components with animations (material inputs, etc.) makes it worse.

#### What is the expected behavior?

Progress spinner has a smooth animation that doesn't shake as it animates.

#### What is the current behavior?

Over time, the progress spinner will begin to shake up and down as it animates.

#### What are the steps to reproduce?

https://stackblitz.com/edit/angular-material2-issue-xg5xgq

1) Load the stackblitz demo
3) Click on the input (trigger the label animation)
4) Click away from the input (return it to normal)
5) Watch for the progress spinner to shake.
6) The progress spinner animation will never recover.

#### What is the use-case or motivation for changing an existing behavior?

The issue appears worse the more material components are on the page. Currently, progress spinner is unusable in our app because the shaking is distracting and unpleasant.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?

Angular: 6.0.3
Material: 6.2.0
Chrome: Version 67.0.3396.79 (Official Build) (64-bit)

**Comment by sahilbabbar** [comment]
I am unable to reproduce this issue. Everything seems fine on my end. Perhaps, something is fishy on the way you are implementing it.

**Comment by josephperrott** [comment]
@HeathJared I am unable to reproduce the issue that you outlined.  Can you confirm if this is still occuring?  Unfortunately we have slightly different version of chrome `67.0.3396.99` so I am not sure if maybe this is the difference.

**Comment by HeathJared** [comment]
Yes, I just double checked today using the provided Stackblitz link. It's easier to see if you zoom the page in.

I have reproduced it in Chrome Version 67.0.3396.99 (Official Build) (64-bit) and in the latest versions of electron.

**Comment by mscud** [comment]
same issue here

**Comment by josephperrott** [comment]
@HeathJared I am still unable to reproduce this.

Are you seeing this in other browsers?  Additionally, what OSes are you seeing this issue on?  Additionally, if you know what GPUs are being used in your reproductions that could prove helpful as well.

**Comment by HeathJared** [comment]
I haven't tried any other browser at the moment.
All of the tests were on Windows 10.

I've seen the issue on the following GPUs:

- GTX 1080
- GTX 1060
- GTX 960M
- GTX 700 series
- Intel HD Graphics 4000
- Intel HD Graphics 530

**Comment by josephperrott** [comment]
@crisbeto Any chance you would be able to take a look in a  Windows 10 environment?

**Comment by crisbeto** [comment]
I was able to reproduce it with @cisasteelersfan, but trying out a few quick solutions didn't help (setting the `transform-origin` in all dimensions, using 3d transforms). Here's an example:

![demo](https://user-images.githubusercontent.com/4450522/43860275-7ade622a-9b53-11e8-9cd9-1d8146b3fece.gif)

**Comment by nilsel** [comment]
Seeing the same "jittery" behaviour, Chrome 68.0.3440.106 on Mac OS 10.13.6. Macbook Pro 2016-model. We're using diameter 20 on the spinners.

However Safari manages to render it flawless, Firefox seems to have minor issues (way more smooth than Chrome, but still seems a bit off).

GPU info:
Radeon Pro 455 2048 MB
Intel HD Graphics 530 1536 MB

**Comment by EdgarSalazar** [comment]
This issue persists to this date.

**Comment by xe4me** [comment]
Still happening for me, it goes away   when I zoom into the page

**Comment by jsessink** [comment]
I've looked into this a bit, and it seems that when the spinners are spinning out of sync with one another, the shakiness goes away.

It's easiest to test with just two - [StackBlitz](https://stackblitz.com/edit/angular-material2-issue-gztkun)

1. Open up dev tools
2. Inspect one of the **circle** elements inside of the svg
3. Un-Check **animation-name** property
4. Re-Check **animation-name** property

They should now be out of sync with one another, and no longer shaking.  Haven't found a fix or workaround for this yet.

**Comment by bbaki** [comment]
We've replaced the material spinner with a one found on codepen:
https://codepen.io/jczimm/details/vEBpoL
MIT licenced (same as all public codepens). Had to do some minor tweaks to it but works just fine for our purposes (watch out as it may not work on IE)

**Comment by tal-abziz** [comment]
if you dont want to use diameter you can work around it using zoom style property
`<mat-spinner style="zoom:0.24"></mat-spinner>`

https://stackblitz.com/edit/angular-material2-issue-h47gmt

**Comment by meta72** [comment]
@tal-abziz sadly, zoom will not work in firefox.
https://www.caniuse.com/#feat=mdn-css_properties_zoom

**Comment by rhartvig** [comment]
If you're looking for a simple fix:

https://stackblitz.com/edit/angular-material2-fallback

Note that this forces spinner into fallback mode. You'll loose the advanced stroke-dashoffset animation, but once Angular Material team fixes it, it's an ease change. This is the gist of it:

```
@Component({
  selector: 'spinner',
  template: `<mat-spinner [diameter]=diameter></mat-spinner>`,
  // when Material team fixes the spinner jitter, remove this provider override
  providers: [{ provide: Platform, useValue: { isBrowser: true, EDGE: true }}]
})
export class Spinner {
  @Input() diameter: number = 100;
}
```
I suspect the problem may be related to https://github.com/angular/components/pull/9833 and the issue: https://github.com/angular/components/issues/8984

**Comment by rhartvig** [comment]
The following seems to work in Chrome, Firefox, Safari and IE:
```
.mat-spinner circle {
  animation-name: progress-spinner-stroke-rotate !important;
  animation-timing-function: linear !important;
}

@keyframes progress-spinner-stroke-rotate {
  $start: 200%; /* approx 2/3 of circle circumference */
  $end: 20%;
  0%   { stroke-dashoffset: $start; }
  50%  { stroke-dashoffset: $end; }
  100% { stroke-dashoffset: $start; transform: rotate(360deg); }
}
```
Test here: https://stackblitz.com/edit/angular-material2-spinner-fix

Removing `transform: rotate(360deg);` causes stroke shaking to reappear. Maybe this is due to outer mat-spinner rotation being countered by inner circle animation keyframes without rotation.

The fix/workaround to avoid stroke shaking is to avoid keyframes without rotation. A full understanding probably requires knowledge of how animating stroke-dashoffset is affected by two rotating elements - maybe someone knows?

Also, the original animation requires special keyframes for every dimension, which makes a general fix hard. Instead I've opted for using percentages (of circle width) to approximate circle circumference, making the above keyframes fit any diameter.

The result is a somewhat different animation but imho still nice and a bit closer to Chrome's own browser tab spinner.

**Comment by rhartvig** [comment]
The following seems to work in Chrome, Firefox, Safari and IE:
```
.mat-progress-spinner-indeterminate-animation {
  /* keep anim slow to avoid wobble from rounding errors */
  animation-duration: 9s !important;
}

.mat-progress-spinner-indeterminate-animation circle {
  animation-name: progress-spinner-stroke-rotate !important;
  animation-timing-function: linear !important;
  animation-duration: 4s !important;
}

@keyframes progress-spinner-stroke-rotate {
  $start: 200%; /* approx 2/3 of circle circumference */
  $end: 20%;
  0%   { stroke-dashoffset: $start; }
  50%  { stroke-dashoffset: $end; }
  /* removing transform: rotate(720deg) causes wobble */
  100% { stroke-dashoffset: $start; transform: rotate(720deg); }
}
```
Test here: https://stackblitz.com/edit/angular-material2-rotate-fix

There are multiple problems addressed:
- The main fix for stroke shaking is to _avoid keyframes without rotation_. Changing `720deg` to `0deg` causes stroke shaking to reappear. Maybe this is due to outer mat-spinner rotation being countered by inner circle animation keyframes without rotation - I'm not sure.
- Fix for small wobble in some browsers from outer rotation, likely due to rounding errors. Since main rotation has been moved to inner element, the outer animation now merely masks the pattern, and thus it can be slowed considerably (10s) to make wobble negligible. 
- While the original animation required special keyframes for every dimension, this solution uses percentages (of circle element width) to calculate circle circumference, making the keyframes fit any diameter. 

The result is a somewhat different animation but imho still nice - and a bit closer to Chrome's own browser tab spinner.

**Comment by BrayoKane** [comment]
Having mat-spinner inside mat-card can assist.

**Comment by blubberbo** [comment]
> .mat-progress-spinner-indeterminate-animation {
>   /* keep anim slow to avoid wobble from rounding errors */
>   animation-duration: 9s !important;
> }
> 
> .mat-progress-spinner-indeterminate-animation circle {
>   animation-name: progress-spinner-stroke-rotate !important;
>   animation-timing-function: linear !important;
>   animation-duration: 4s !important;
> }
> 
> @keyframes progress-spinner-stroke-rotate {
>   $start: 200%; /* approx 2/3 of circle circumference */
>   $end: 20%;
>   0%   { stroke-dashoffset: $start; }
>   50%  { stroke-dashoffset: $end; }
>   /* removing transform: rotate(720deg) causes wobble */
>   100% { stroke-dashoffset: $start; transform: rotate(720deg); }
> }

I tried this but I get the same behavior. To clarify - do you leave the `diameter` property on, or remove it?

**Comment by rhartvig** [comment]
@blubberbo please check the stackblitz link. Make sure to set ViewEncapsulation.None for styles file or place styles in global styles.scss. Diameter is set. Let me know if demo above works for you.

**Comment by blubberbo** [comment]
> @blubberbo please check the stackblitz link. Make sure to set ViewEncapsulation.None for styles file or place styles in global styles.scss. Diameter is set. Let me know if demo above works for you.

placing them in the global styles.scss file fixed it, you are a gentleman and a scholar!

**Comment by Ploppy3** [comment]
@blubberbo I don't get it, your solution is actually breaking the animation, yes it does not flicker, but it does not look like the original animation anymore.

I think it has to be said.

**Comment by blubberbo** [comment]
> @rhartvig I don't get it, your solution is actually breaking the animation, yes it does not flicker, but it does not look like the original animation anymore.
> 
> I think it has to be said.

this is true, it is not the same animation as the original, but not flickering is more important to me. Ideally, it would not flicker and be the original animation. This solution fixed my primary purpose but not a complete fix to the issue

**Comment by Mike--S** [comment]
Hello guys, I know it's not always convenient, but you can just hide spinner and show it using opacity property, instead of display property or ngIf, it seems like if you don't change the flow for this component, then it works well
https://stackblitz.com/edit/angular-material2-rotate-fix-uv6ggd

**Comment by phani17c** [comment]
I have commented animation of mat-spinner styles, it is okay.
